Support private Route53 zones

JonRoma opened the pull request:
*   Module now supports private Route53 zones. In this case, the A record is set to the private rather than public IP.

*   One can now set "associate_public_ip_address" to false for instances that don't have public Route53 presence.

These discoveries were made in the course of testing networking for a new VPC.

JonRoma commented:
@ddriddle, would you please be so kind as to give this a review?

The logic to compute the IP address based on whether an elastic IP address is selected could probably be simplified, particularly in the case of a Route 53 alias in a private zone, where EIP makes no sense. In theory specifying EIP and private_zone should produce an error, but I didn't bother with that.

ddriddle commented on .github/workflows/terraform.yml:
```diff
@@ -0,0 +1,81 @@
+
+on:
+  push:
+#   branches:
```
Please, drop the unnecessary comments on lines 7 and 8.

ddriddle commented on .github/workflows/terraform.yml:
```diff
@@ -0,0 +1,81 @@
+# Some of this logic was stolen from https://www.freecodecamp.org/news/a-lightweight-tool-agnostic-ci-cd-flow-with-github-actions/
```
You have made so many changes I don't think you need to credit this link anymore. IMO.

JonRoma replied:
@ddriddle, I should have asked you not to worry about reviewing the GitHub actions. The file here is a minimally-viable version borrowed from other Terraform projects, and as such does not represent anything new to this repo, nor does it represent what I hope to attain as a gold standard for GitHub Actions for Terraform modules.

This is one of the areas that I hope to work on with you in the future after Shibboleth has died down. To that end, I'm going to ignore the comments on the CI pieces, with the thought that we'll address them together soon, add them to SDG's templates, and disperse them broadly across all our repos so that we can stick a fork in Drone.

ddriddle commented on .github/workflows/terraform.yml:
```diff
@@ -0,0 +1,81 @@
+          terraform_version: 0.12.29
+      #   cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }}
+
+#     TODO: This step duplicates work done by the Makefile.
```
Please, remove this comment block (lines 27-31)

ddriddle commented on .github/workflows/terraform.yml:
```diff
@@ -0,0 +1,81 @@
+        run: |
+          make all
+
+#     - name: terraform init
```
Please remove this unused code (lines 39-81).

ddriddle commented on route53.tf:
```diff
@@ -1,15 +1,28 @@
   count = length(var.alias)
 
-  name = lookup(var.alias[count.index], "domain", "")
+  name         = local.alias[count.index].domain
```
@JonRoma What is the advantage of using the `local.alias` compared to just using a straight `lookup` as was done before?

JonRoma replied:
The purpose of the added local variable was to facilitate the addition of the VPC id for private zones in a way that is reasonably readable. Perhaps there are other ways to accomplish that – do you have ideas?

While creating a bunch of intermediate variables sometimes annoys me, it's something I had to do heavily in the `enterprise-subnets` module to be generic and to use sophisticated data structures without making the code a complete nightmare of nested maps and lists that no one including me will be able to read three months from now. :-)

I don't think sheer efficiency [by avoiding temporary variables] is worth fretting about in Terraform code, and I tend to err on the side of readability vs. efficiency.

ddriddle replied:
I'm not worried about efficiency here, my only concern is that the diff is larger than necessary which does make it harder to review and understand why the changes are happening. Other than that your approach is fine. I don't think it improves readability much here, but in general I agree it is a good idea, and much better than spreading the logic all over the place.

> The purpose of the added local variable was to facilitate the addition of the VPC id for private zones in a way that is reasonably readable. Perhaps there are other ways to accomplish that – do you have ideas?

What I would have done is simply inlined the `vpc_id` logic since it is only used in one place. If the logic was needed in more than one place then I would have gone with the same approach you went with. Here is how I would refactor the code:

```terraform
data "aws_route53_zone" "selected" {
  count = length(var.alias)
  
  name         = lookup(var.alias[count.index], "domain", "")
  private_zone = lookup(var.alias[count.index], "private_zone", false)
  vpc_id       = lookup(var.alias[count.index], "private_zone", false) ? data.aws_vpc.selected.id : null
}

resource "aws_route53_record" "default" {
  count = length(var.alias)

  zone_id = element(data.aws_route53_zone.selected.*.zone_id, count.index)
  name    = lookup(var.alias[count.index], "hostname", var.name)
  type    = "A"
  ttl     = lookup(var.alias[count.index], "ttl", 60)
  records = [lookup(var.alias[count.index], "private_zone", false) ? local.private_ip : local.public_ip]
}
```

I'm fine keeping the object or refactoring it out. Your call. Thanks for your comments!

JonRoma replied:
@ddriddle, honestly this is why I like reviewing each other's code. I added the logic I did because at first glance, the addition of private zone support looked more complicated than it actually was. And when I finished, I was tired enough that I forgot to take a step back and see that the extra data structure wasn't needed.

I've refactored the `local.alias` out of the module.

ddriddle commented on route53.tf:
```diff
@@ -1,15 +1,28 @@
+locals {
+  alias = [for a in var.alias : {
+    domain       = a.domain
+    hostname     = a.hostname
```
Before if the hostname was not specified it defaulted to `var.name`. Did you intend to change that behavior? It is documented in the README and the code for that use to be on line 11 below.

ddriddle replied:
Please change line 4 to:

```
hostname = lookup(a, "hostname", var.name)
```

JonRoma replied:
This was **not** intentional; good catch. I have fixed it in the latest push. 

Please note that, if the user specifies more than one alias but omits the hostname argument on both, Route 53 will reject the `apply` because of a duplicate hostname. That's obviously a user error, however.

ddriddle reviewed (changes requested):
@JonRoma I think the code for calculating the private IP and DNS is good, as it allows for the code to evolve in a way that we can be sure that the values returned are always correct. I agree, you could simplify the code but I prefer the code as is since it is more defensive.

I found a bug. The hostname should default to `var.name` if not specified. That logic was dropped. If you intended that change then you need to update the documentation, otherwise please apply the one line change I suggested.

Overall the code looks good, but I wonder if `local.alias` makes the code more complicated. I'm not sure. You may want to consider refactoring it out. There may well be a very good reason for the object that I am not seeing.

JonRoma commented:
@ddriddle, I have addressed the issue you pointed out, and have commented on the others. If you are satisfied, please give an approving review so that this can go forward. Otherwise, let's talk. :-)

ddriddle commented on route53.tf:
```diff
@@ -1,15 +1,28 @@
 }
 
 resource "aws_route53_record" "default" {
   count = length(var.alias)
```
I think it would be better to change this to:

```
count = length(local.alias)
```

JonRoma replied:
This is now moot due to my refactoring, but I agree that the count should match the variable being iterated over, be it in `for_each` or with `count.index`.

As an aside, I found an interesting tidbit in that Terraform's `keys` and `values` function always returns a list that's in lexicographical order, and is hence deterministic. I have used this to my advantage elsewhere. :-)

ddriddle commented on route53.tf:
```diff
@@ -1,15 +1,28 @@
+}
+
 data "aws_route53_zone" "selected" {
   count = length(var.alias)
```
I think it would be better to directly count `local.alias` since `var.alias` is not directly used here anymore:

```
count = length(local.alias)
```

ddriddle reviewed (changes requested):
Please change the count lines if you choose not to remove `local.alias`. Thanks for asking for my review and your thoughtful comments. Enjoy the snow :-D

JonRoma commented:
@ddriddle, the refactored code is ready for your review.

Also, at a customer request, I added the hostname to the list of template variables passed in to the `user_data` script. I had to derive it from the first element of the `alias` variable. A more logical way would have been to use the value from `as_aws_instance.fqdn`, but that won't work because that value isn't known until the instance is created – a chicken *vs.* egg problem!

ddriddle commented on data.tf:
```diff
@@ -67,12 +67,15 @@ locals {
   efs_file_system_id = lookup(var.efs_file_system, "file_system_id", "")
   efs_mount_point    = lookup(var.efs_file_system, "mount_point", "")
   efs_source_path    = lookup(var.efs_file_system, "source_path", "")
+  hostname           = length(var.alias) > 0 ? format("%s.%s", var.alias[0].hostname, var.alias[0].domain) : null
```
Line 70 has a bug: hostname is optional, and should default to `var.name` per the docs, but here it is treated like a required variable, so it should be:

```terraform
length(var.alias) > 0 ? format("%s.%s", lookup(var.alias[0], "hostname", var.name), var.alias[0].domain) : null
```

JonRoma replied:
Code has been refactored per your suggestions. 

To be completely honest, a user omitting something as critical as a hostname or domain from the `alias` block is probably being foolish.

ddriddle commented on route53.tf:
```diff
@@ -1,15 +1,17 @@
 
   zone_id = element(data.aws_route53_zone.selected.*.zone_id, count.index)
-  name    = lookup(var.alias[count.index], "hostname", var.name)
+  name    = var.alias[count.index].hostname
```
Please revert this change, otherwise update the documentation to indicate that `var.name` is not the default.

ddriddle reviewed (changes requested):
The documentation still indicates that `var.name` is the default value for hostname in an alias block. The code in this pull request does not implement that feature. I made suggestions in the comments below on how to easily add that feature back. If the feature is no longer desirable then please update the documentation.

JonRoma committed:
Support private Route53 zones

*   Module now supports private Route53 zones. In this case, the A record
    is set to the private rather than public IP.

*   One can now set "associate_public_ip_address" to false for instances
    that don't have public Route53 presence.

*   Switch module from Drone to GitHub Actions.

*   Add `hostname` to list of variables passed to template used by the
    user_data script.

JonRoma commented:
@ddriddle, please take another look.

ddriddle reviewed (approved):
Everything looks good!
